### DISCUSSION TITLE
Question about Performance

### DISCUSSION THREAD

**Discussion by anandsork** [chatwoot/chatwoot#1358] [Self hosted]
This is not an issue, but a question.
 How is the performance of the system for self-hosted solution? Can it handle 1000s of users at a time? The messages will reach to millions within 6 months of time. Any specific OS configuration needed for server & DB?

**Comment by sojan-official** [comment]
@anandsork, we handle a decent amount of traffic in our cloud version, which runs using the same code base.  The performance aspect is mostly based on how you allocate the resources as your traffic starts growing.  

Initially, you could keep adding web/workers as the traffic increases. If you are facing a bottleneck at the database side, you could allocate more resources there.  When you run into cases where you max out the resources you can allocate for a single server, you could look into implementing replicas to handle the reads and so ... 

So, Inherently there aren't any limits in the code that's stopping you from handling the scale. It's more about tweaking the architecture where chatwoot is running based on your growing requirements.  

If you face any specific issues, our team will be happy to help in resolving them. We also offer premium support solutions, where our engineers assist you with your self-hosted solutions, if that's something that interests you.

**Comment by anandsork** [comment]
Yes. Horizontal scaling is possible. My question is more towards the limits it can reach to (or break) on a single server before we scale. Say 4GB  memory machines (DB is on one instance and Other instance hold web+worker).
For costs, we always try to optimize the resources. So its good to know some kind of limits.

**Reply by sojan-official** [reply]
Makes sense. Unfortunately, we still need to do more performance testing in that aspect. So I don't have much metrics to share for the moment.  
Would be happy to lend a hand if someone from the community decides to volunteer.

**Comment by fiantinangon** [comment]
Hi @sojan-official related with this question, I've done performance test to send 1000 message from 10 diff numbers, and the chatwoot performance seems so slow while open the conversation list, cause chatwoot download all the conversation data which is thousand of chat data on that 10 conversations, why not load the chat when selecting the conversation ? I attached the snapshot about the requested timeframe btw

<img width="1280" alt="Screen Shot 2020-12-09 at 9 00 38 PM" src="https://user-images.githubusercontent.com/9161278/101639947-7e978f80-3a62-11eb-9ef4-6e393ad9dca7.png">
<img width="1280" alt="Screen Shot 2020-12-09 at 8 58 46 PM" src="https://user-images.githubusercontent.com/9161278/101639968-85260700-3a62-11eb-9e56-6c24ec7d13dd.png">


and this is the sample response data when loading the conversation list, and it's surprisingly big
<img width="408" alt="Screen Shot 2020-12-09 at 9 07 43 PM" src="https://user-images.githubusercontent.com/9161278/101640100-a850b680-3a62-11eb-9b74-1242d20629ea.png">

**Reply by pranavrajs** [reply]
@fiantinangon What is the server spec you are using? I don't think we are list all the messages at once. Only the last message should be listed.

**Reply by fiantinangon** [reply]
This is the spec @pranavrajs  :
CPU: 2x2.6GHz Intel® Xeon® Gold 6132
RAM: 8 GB DDR4-2666 ECC REG
Storage: 80 GB SSD

The condition is we have 10 latest conversations each conversation have 100 unread messages, and when chatwoot perform request to get conversation list with the above condition it takes to long, sometimes need more than 20s to download the response as I attached above

And if I read all the unread conversation and trying to get the conversations list it performs faster than before, you can see on the image bellow
<img width="1280" alt="Screen Shot 2020-12-09 at 9 31 28 PM" src="https://user-images.githubusercontent.com/9161278/101642927-2a8eaa00-3a66-11eb-8fb8-45d3ff83898e.png">

**Reply by pranavrajs** [reply]
@fiantinangon Let me check this out, will see if there is an immediate fix.

**Reply by fiantinangon** [reply]
> @fiantinangon Let me check this out, will see if there is an immediate fix.

Ok @pranavrajs let me know if you need any more data to help reproduce this issue, looking forward for the update

**Reply by pranavrajs** [reply]
@fiantinangon Can you please pull the latest develop branch and see if this is still an issue?

**Reply by fiantinangon** [reply]
@pranavrajs Ok, i'll test it asap

**Reply by fiantinangon** [reply]
Hi @pranavrajs i got this error after switch to develop branch and do

# Update dependencies
bundle
yarn

# Recompile the assets
rake assets:precompile RAILS_ENV=production

# Migrate the database schema
RAILS_ENV=production bundle exec rake db:migrate

# Restart the chatwoot server
systemctl restart chatwoot.target

<img width="648" alt="Screen Shot 2020-12-12 at 6 41 24 AM" src="https://user-images.githubusercontent.com/9161278/101964981-c0ccf680-3c45-11eb-924d-9731aee32e9c.png">
<img width="646" alt="Screen Shot 2020-12-12 at 6 56 39 AM" src="https://user-images.githubusercontent.com/9161278/101965440-44d3ae00-3c47-11eb-8215-ea8b954314d7.png">


I see on the log the issue related with web pack 

```
Dec 12 08:04:58 chatwoot-web.1[498230]: I, [2020-12-12T08:04:58.829296 #498230]  INFO -- : [d4c91f18-b725-425f-9838-eb0f2ff022ea] Started GET "/app/login" for 125.163.233.247 at 2020-12-12 08:04:58 +0200
Dec 12 08:04:58 chatwoot-web.1[498230]: I, [2020-12-12T08:04:58.830442 #498230]  INFO -- : [d4c91f18-b725-425f-9838-eb0f2ff022ea] Processing by DashboardController#index as HTML
Dec 12 08:04:58 chatwoot-web.1[498230]: I, [2020-12-12T08:04:58.830504 #498230]  INFO -- : [d4c91f18-b725-425f-9838-eb0f2ff022ea]   Parameters: {"params"=>"login"}
Dec 12 08:04:58 chatwoot-web.1[498230]: I, [2020-12-12T08:04:58.832617 #498230]  INFO -- : [d4c91f18-b725-425f-9838-eb0f2ff022ea]   Rendering dashboard/index.html.erb within layouts/vueapp
Dec 12 08:04:58 chatwoot-web.1[498230]: I, [2020-12-12T08:04:58.832760 #498230]  INFO -- : [d4c91f18-b725-425f-9838-eb0f2ff022ea]   Rendered dashboard/index.html.erb within layouts/vueapp (Duration: 0.1ms | Allocations: 4)
Dec 12 08:04:58 chatwoot-web.1[498230]: I, [2020-12-12T08:04:58.835194 #498230]  INFO -- : [d4c91f18-b725-425f-9838-eb0f2ff022ea] Completed 500 Internal Server Error in 5ms (Allocations: 2764)
Dec 12 08:04:58 chatwoot-web.1[498230]: F, [2020-12-12T08:04:58.836082 #498230] FATAL -- : [d4c91f18-b725-425f-9838-eb0f2ff022ea]
Dec 12 08:04:58 chatwoot-web.1[498230]: [d4c91f18-b725-425f-9838-eb0f2ff022ea] ActionView::Template::Error (Webpacker can't find application.css in /home/chatwoot/chatwoot/public/packs/manifest.json. Possible causes:
Dec 12 08:04:58 chatwoot-web.1[498230]: 1. You want to set webpacker.yml value of compile to true for your environment
Dec 12 08:04:58 chatwoot-web.1[498230]:    unless you are using the `webpack -w` or the webpack-dev-server.
Dec 12 08:04:58 chatwoot-web.1[498230]: 2. webpack has not yet re-run to reflect updates.
Dec 12 08:04:58 chatwoot-web.1[498230]: 3. You have misconfigured Webpacker's config/webpacker.yml file.
Dec 12 08:04:58 chatwoot-web.1[498230]: 4. Your webpack configuration is not creating a manifest.
Dec 12 08:04:58 chatwoot-web.1[498230]: Your manifest contains:
Dec 12 08:04:58 chatwoot-web.1[498230]: {
Dec 12 08:04:58 chatwoot-web.1[498230]:   "application.js": "/packs/js/application-f9836274d54f9e6405e1.js",
Dec 12 08:04:58 chatwoot-web.1[498230]:   "application.js.map": "/packs/js/application-f9836274d54f9e6405e1.js.map",
Dec 12 08:04:58 chatwoot-web.1[498230]:   "entrypoints": {
Dec 12 08:04:58 chatwoot-web.1[498230]:     "application": {
Dec 12 08:04:58 chatwoot-web.1[498230]:       "js": [
Dec 12 08:04:58 chatwoot-web.1[498230]:         "/packs/js/application-f9836274d54f9e6405e1.js"
Dec 12 08:04:58 chatwoot-web.1[498230]:       ],
Dec 12 08:04:58 chatwoot-web.1[498230]:       "js.map": [
Dec 12 08:04:58 chatwoot-web.1[498230]:         "/packs/js/application-f9836274d54f9e6405e1.js.map"
Dec 12 08:04:58 chatwoot-web.1[498230]:       ]
Dec 12 08:04:58 chatwoot-web.1[498230]:     },
Dec 12 08:04:58 chatwoot-web.1[498230]:     "sdk": {
Dec 12 08:04:58 chatwoot-web.1[498230]:       "js": [
Dec 12 08:04:58 chatwoot-web.1[498230]:         "/packs/js/sdk.js"
Dec 12 08:04:58 chatwoot-web.1[498230]:       ],
Dec 12 08:04:58 chatwoot-web.1[498230]:       "js.map": [
Dec 12 08:04:58 chatwoot-web.1[498230]:         "/packs/js/sdk.js.map"
Dec 12 08:04:58 chatwoot-web.1[498230]:       ]
Dec 12 08:04:58 chatwoot-web.1[498230]:     },
Dec 12 08:04:58 chatwoot-web.1[498230]:     "superadmin_pages": {
Dec 12 08:04:58 chatwoot-web.1[498230]:       "js": [
Dec 12 08:04:58 chatwoot-web.1[498230]:         "/packs/js/superadmin_pages-f9836274d54f9e6405e1.js"
Dec 12 08:04:58 chatwoot-web.1[498230]:       ],
Dec 12 08:04:58 chatwoot-web.1[498230]:       "js.map": [
Dec 12 08:04:58 chatwoot-web.1[498230]:         "/packs/js/superadmin_pages-f9836274d54f9e6405e1.js.map"
Dec 12 08:04:58 chatwoot-web.1[498230]:       ]
Dec 12 08:04:58 chatwoot-web.1[498230]:     },
Dec 12 08:04:58 chatwoot-web.1[498230]:     "superadmin": {
Dec 12 08:04:58 chatwoot-web.1[498230]:       "js": [
Dec 12 08:04:58 chatwoot-web.1[498230]:         "/packs/js/superadmin-f9836274d54f9e6405e1.js"
Dec 12 08:04:58 chatwoot-web.1[498230]:       ],
Dec 12 08:04:58 chatwoot-web.1[498230]:       "js.map": [
Dec 12 08:04:58 chatwoot-web.1[498230]:         "/packs/js/superadmin-f9836274d54f9e6405e1.js.map"
Dec 12 08:04:58 chatwoot-web.1[498230]:       ]
Dec 12 08:04:58 chatwoot-web.1[498230]:     },
Dec 12 08:04:58 chatwoot-web.1[498230]:     "widget": {
Dec 12 08:04:58 chatwoot-web.1[498230]:       "js": [
Dec 12 08:04:58 chatwoot-web.1[498230]:         "/packs/js/widget-f9836274d54f9e6405e1.js"
Dec 12 08:04:58 chatwoot-web.1[498230]:       ],
Dec 12 08:04:58 chatwoot-web.1[498230]:       "js.map": [
Dec 12 08:04:58 chatwoot-web.1[498230]:         "/packs/js/widget-f9836274d54f9e6405e1.js.map"
Dec 12 08:04:58 chatwoot-web.1[498230]:       ]
Dec 12 08:04:58 chatwoot-web.1[498230]:     }
Dec 12 08:04:58 chatwoot-web.1[498230]:   },

```
Is there any special config we need to compile `develop` branch @pranavrajs ?

but if I running it using `rails server` manually it's work hmm what's wrong with the chatwoot.target ?
<img width="1146" alt="Screen Shot 2020-12-12 at 7 01 47 AM" src="https://user-images.githubusercontent.com/9161278/101965682-202c0600-3c48-11eb-8689-68be1320dec9.png">
<img width="1280" alt="Screen Shot 2020-12-12 at 7 02 02 AM" src="https://user-images.githubusercontent.com/9161278/101965690-24582380-3c48-11eb-878a-3fb1e7081cfd.png">

**Reply by pranavrajs** [reply]
Please try this.
```
bundle
yarn
rake assets:clobber assets:precompile RAILS_ENV=production
```

```
RAILS_ENV=production bundle exec rake db:migrate

cp /home/chatwoot/chatwoot/deployment/chatwoot-web.1.service /etc/systemd/system/chatwoot-web.1.service
cp /home/chatwoot/chatwoot/deployment/chatwoot-worker.1.service /etc/systemd/system/chatwoot-worker.1.service
cp /home/chatwoot/chatwoot/deployment/chatwoot.target /etc/systemd/system/chatwoot.target

systemctl restart chatwoot.target
```

**Reply by fiantinangon** [reply]
Thanks @pranavrajs thanks it works, i spot the diff is on this line `rake assets:clobber assets:precompile RAILS_ENV=production` so I guess we need to wipe all the compiled assets after checkout to other branch or doing upgrade, maybe we need to update the documentation upgrade script

**Reply by pranavrajs** [reply]
@fiantinangon Precompile issues do not happen always. That's why we didn't add it to the upgrade steps. I will add it to the documentation.

**Reply by fiantinangon** [reply]
Ok @pranavrajs i'll continue test the unread conversation performance and i'll get back asap about the result

**Comment by sojan-official** [comment]
We will be putting more efforts into load testing our critical endpoints using locust. 
Follow the work at https://github.com/chatwoot/chatwoot-locust-tests

**Comment by NusretOzates** [comment]
Hello! If I understand this right, we can safely increase the pod count of both web and worker servers, right?

**Reply by sojan-official** [reply]
That is correct. Just ensure that the database and Redis configs provided to these pods are the same.

**Reply by NusretOzates** [reply]
Awesome, thanks!
